### PR TITLE
[SPARK-34220][ML] BucketedRandomProjectionLSH transform optimization

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
use GEMV instead of DOT

### Why are the changes needed?
1, better performance, could be 20% faster than existing impl
2, simplify model saving

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
existing testsuites

### CODE REVIEW

**Comment by zhengruifeng** [comment]
test code:
```
import org.apache.spark.ml.linalg._
import org.apache.spark.ml.feature._
import org.apache.spark.storage.StorageLevel


val df = spark.read.format("libsvm").load("/d1/Datasets/epsilon/epsilon_normalized.t")
df.persist(StorageLevel.MEMORY_AND_DISK)
df.count



val brp = new BucketedRandomProjectionLSH().setNumHashTables(100).setInputCol("features").setOutputCol("projected").setBucketLength(1.0).setSeed(12345)

val brpModel = brp.fit(df)
Seq.range(0, 1000).foreach { i => brpModel.transform(df).count } // warm up


val start = System.currentTimeMillis; Seq.range(0, 1000).foreach { i => brpModel.transform(df).count }; val end = System.currentTimeMillis; val duration = end - start; 
```

master:
start: Long = 1611547911926
end: Long = 1611547940855
duration: Long = 28929





this PR:
start: Long = 1611548225365
end: Long = 1611548249266
duration: Long = 23901

File: mllib/src/main/scala/org/apache/spark/ml/feature/BucketedRandomProjectionLSH.scala [hunk 1]
```diff
@@ -60,14 +59,20 @@ private[ml] trait BucketedRandomProjectionLSHParams extends Params {
...
     override val uid: String,
-    private[ml] val randUnitVectors: Array[Vector])
+    private[ml] val randMatrix: Matrix)
   extends LSHModel[BucketedRandomProjectionLSHModel] with BucketedRandomProjectionLSHParams {
 
+  private[ml] def this(uid: String, randUnitVectors: Array[Vector]) = {
+    this(uid, Matrices.fromVectors(randUnitVectors))
+  }
+
+  private[ml] def randUnitVectors: Array[Vector] = randMatrix.rowIter.toArray
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/feature/BucketedRandomProjectionLSH.scala:74]: add this constructor and method `randUnitVectors` for existing testsuites

**Comment by zhengruifeng** [comment]
friendly ping @srowen

**Review by srowen — APPROVED** [review]

**Comment by zhengruifeng** [comment]
Merged to master, thanks @srowen for reivewing!

**Pull request closed without merging**